Add `isRequired` and `isOptional` to form fields

alex-ju opened the pull request:
### :pushpin: Summary

Add `isRequired` and `isOptional` to form fields.

### :hammer_and_wrench: Detailed description

Following the recommendations, we add these options to the label/legend element and then we expose them via field/fieldset to form components.

This will allow components to be called as follows:

```hbs
<Hds::Form::TextInput::Field @isRequired={{true}} as |F|>
  <F.Label>This is the label text</F.Label>
</Hds::Form::TextInput::Field>
```

An alternative would be to preserve `isRequired` only at the label/legend level, removing the need for most of the code in this PR, but surfacing this argument at the field level brings it in-line with the `isInvalid` argument, so I guess it makes sense to have them aligned.
```hbs
<Hds::Form::TextInput::Field as |F|>
  <F.Label @isRequired={{true}}>This is the label text</F.Label>
</Hds::Form::TextInput::Field>
```

**Note**
I omitted the toggle component as I've not seen a reference in design and don't think it makes sense for the component to have it, but open to discussions. Similarly for checkbox/radio as a standalone field.

### :link: External links

[Jira issue](https://hashicorp.atlassian.net/jira/software/c/projects/HDS/boards/445?modal=detail&selectedIssue=HDS-137)

***

### 👀 How to review

👉 Review commit-by-commit

Reviewer's checklist:

- [ ] +1 Percy if applicable

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

alex-ju commented:
@heatherlarsen if you'd like to have a look and/or provide feedback at this stage (as you know, the form components are not yet accessible from the index page, but URLs are https://hds-components-git-alex-ju-form-fields-require-8a8dd0-hashicorp.vercel.app/components/form/text-input (change `text-input` with `textarea` or `select`)

didoo commented on packages/components/tests/dummy/app/templates/components/form/base-elements.hbs:
```diff
@@ -50,6 +50,16 @@
         element.</p>
     </dd>
+    <dt>isRequired <code>boolean</code></dt>
+    <dd>
+      <p>It appends a badge to the label text to indicate user input is required.</p>
+      <p>Default: <span class="default">false</span></p>
+    </dd>
+    <dt>isOptional <code>boolean</code></dt>
+    <dd>
+      <p>It appends "(Optional)" to the label text to indicate user input is not required.</p>
+      <p>Default: <span class="default">false</span></p>
+    </dd>
```
[nitpick] Probably I would generalize the text, so that we don't have to specify there is some kind of difference between`badge` and `("Optional")` (it's kinda an internal thing).  Something like this:
- It appends a [mark|indication] to the label text to indicate that the user input is required.
- It appends a [mark|indication] to the label text to indicate that the user input is optional.

didoo commented on packages/components/tests/dummy/app/templates/components/form/base-elements.hbs:
```diff
@@ -406,6 +446,40 @@
+  />
+  {{! template-lint-enable no-unbalanced-curlies }}
+  {{! prettier-ignore-end }}
+  <p class="dummy-paragraph">Renders to:</p>
```
[suggestion] probably, now that we have the `Required` badge, I would update the example below this (the one with "Some badge") to have the badge as `highlight` color, so they are different and it doesn't generate confusion (see all the occurrences of "Some badge" in the page).

alex-ju replied:
Updated with the suggestion – have thought about this, but hesitated as people may take it as a recommendation to use the a colored badge in label.

didoo commented on packages/components/tests/dummy/app/templates/components/form/base-elements.hbs:
```diff
@@ -905,19 +1013,31 @@
   <Hds::Form::Label>
     <div class="dummy-form-base-elements-label-with-badge">This is the label
-      <Hds::Badge @size="small" @text="Some badge" /></div>
+      <Hds::Badge @size="small" @text="Some badge" />
```
[question/suggestion] do you think we should showcase an example of "With structured content + Required" and/or "With structured content + Optional" (probably with short text and with long text)? On one side would be nice so we can test it works as expected, but on the other I fear it would confuse the end consumers...

alex-ju replied:
Seeing Heather backed this suggestion I added a few more examples.

didoo commented:
> I omitted the toggle component as I've not seen a reference in design and don't think it makes sense for the component to have it

@alex-ju probably for the same reason for which we re-introduced the "group" variant for the toggle, and we showcase it as multiple toggles even if likely is not a real use case, I would be inclined to add it anyway (maybe a user need to turn on a "toggle" for some reasons, like accepting the risk of an operation, or explicitly approving something)

didoo reviewed (commented):
Overall code is OK for me. There are a few things I would like to discuss with you (and also with @heatherlarsen / @cveigt about the position of the "required/optional" indicators) before approving the PR.

MelSumner commented on packages/components/addon/components/hds/form/field/index.js:
```diff
@@ -10,6 +10,8 @@ export const LAYOUT_TYPES = ['vertical', 'flag'];
 export default class HdsFormFieldIndexComponent extends Component {
   @tracked ariaDescribedBy = getAriaDescribedBy(this);
   @tracked descriptors = [];
+  @tracked isRequired = this.args.isRequired;
```
[question] I might be missing something here, but why is this tracked?

MelSumner commented on packages/components/addon/components/hds/form/field/index.js:
```diff
@@ -10,6 +10,8 @@ export const LAYOUT_TYPES = ['vertical', 'flag'];
 export default class HdsFormFieldIndexComponent extends Component {
   @tracked ariaDescribedBy = getAriaDescribedBy(this);
   @tracked descriptors = [];
+  @tracked isRequired = this.args.isRequired;
```
[issue] I don't see where this is also adding the `required` attribute to the input element if true.

MelSumner replied:
If we are providing a way for the author to say "isRequired" - and that provides a required badge, do you think it would be logical to also add the `required` attribute to the associated `input` element?

didoo replied:
Since we've decided to leave responsibility to the consumers to handle most part of the HTML attributes for the controls, especially validation, I think we should follow the same approach to `required` here: if a field is required and the consumers handle if and how to validate that the field is "filled" by the user, if and how to show an error, then adding a `required` attribute would go against this approach, potentially may be unexpected too (we may have client-side validation interfering with the consumers-validation, at UI level).

alex-ju replied:
Sure, I'll get ambassadors' view on this.

MelSumner commented on packages/components/tests/dummy/app/templates/components/form/base-elements.hbs:
```diff
@@ -50,6 +50,16 @@
+    <dt>isRequired <code>boolean</code></dt>
+    <dd>
+      <p>It appends a badge to the label text to indicate user input is required.</p>
+      <p>Default: <span class="default">false</span></p>
```
[issue] if you are leaving this up to the user to declare, then the default would be `null` instead of `false`, yes?

didoo replied:
Technically correct. But probably, to keep it simple, we can leave it as is? For the consumer what is important is to know is that it can be `true/false` (and in this case the default would be _falsy_, but if we use `null` or `falsy` may be confusing?)

didoo commented:
@heatherlarsen @cveigt + @alex-ju something I want to discuss with you, as a possible alternative to the current layout implementation of the "required/optional" indicators (currently inline with the text). Let me show you the problem. 

Let's take the default example:
<img width="325" alt="screenshot_1612" src="https://user-images.githubusercontent.com/686239/177410485-094ba59d-66ee-4fc1-b247-06e2c2c70760.png">
this works fine, no problem here. But what if the text is slightly longer? well in some cases (it depends by the length of the text and the single words) this could happen:
<img width="281" alt="screenshot_1613" src="https://user-images.githubusercontent.com/686239/177410683-a85a7b83-5d9c-4354-a55e-0ed315b304a0.png">
Essentially, being an inline element, and there not being enough space after the text, the mark goes on a new line. Not only, it has a 4px margin from the left, because... that's what we declared to have, to avoid it touching the text when is next to a word.

Now, not a big problem per se, still works, but I want to suggest an alternative. To have the label text and the mark laid out via a `flex` container, so that this is the alternative behaviour when there is not enough space:
<img width="279" alt="screenshot_1614" src="https://user-images.githubusercontent.com/686239/177414649-4040ae12-5ee9-40f0-a1c4-017be80bb762.png">
Of course, in case _there is_ enough space, the layout would be visually indistinguishable from the current one:
<img width="277" alt="screenshot_1615" src="https://user-images.githubusercontent.com/686239/177415072-672f8edd-2a48-40ff-b86c-a0c46164a2dd.png">

Essentially I am suggesting to go from this:
![inline](https://user-images.githubusercontent.com/686239/177416075-92ee5911-fccf-4a73-888e-1baf771cec6e.png)
to this:
![flex](https://user-images.githubusercontent.com/686239/177416433-07839658-6810-4650-b35c-dd7290590963.png)

This, on the Figma side, would also solve the problem you discussed a few days ago, about the layout with "required/optional" not being responsive on resize, while with the proposed change this could be achieved with autolayout.

What do you think?

heatherlarsen commented:
@didoo While it's harder to accomplish in Figma (and would require a detach of the component), the indicator wrapping exactly to that point is such an edge case that I don't think we should be adjusting our whole approach around it. Personally, the inline approach feels like it provides a more natural reading/scanning experience.

Here is an example of a label from our products that would result in a very unnatural reading experience with the proposed change: 

<img width="303" alt="Frame 3255" src="https://user-images.githubusercontent.com/8553306/177419464-f5514bd3-3486-4247-94b9-8f5143541e4a.png">


[question] Could you move the 4px margin to the text and not the badge so that the gap on the left is avoided?

heatherlarsen commented on packages/components/app/styles/components/badge.scss:
```diff
@@ -144,3 +144,12 @@ $size-props: (
+  .hds-badge {
+    display: inline-block;
+    vertical-align: baseline;
+    margin-left: 4px;
```
[nitpick(ux)] font-weight is explicitly called out for the optional indicator, but for required it's taking on the font-weight from the label. The font-weight of the required badge should be `--token-typography-font-weight-medium`.

cveigt replied:
I agree with Heather here. We may also have more indicators beyond required and optional with longer text, which would make the label even more challenging to read.

didoo replied:
@heatherlarsen the problem is because we (I) never explicitly set the `font-weight` on the `Badge` component.
I've opened a PR to fix this: https://github.com/hashicorp/design-system/pull/426
/cc @alex-ju

cveigt commented:
I agree with Heather here. We may also have more indicators beyond required and optional with longer text, which would make the label even more challenging to read.

alex-ju commented on packages/components/addon/components/hds/form/field/index.js:
```diff
@@ -10,6 +10,8 @@ export const LAYOUT_TYPES = ['vertical', 'flag'];
 export default class HdsFormFieldIndexComponent extends Component {
   @tracked ariaDescribedBy = getAriaDescribedBy(this);
   @tracked descriptors = [];
+  @tracked isRequired = this.args.isRequired;
+  @tracked isOptional = this.args.isOptional;
```
We don't need a setter on these values, but we at least need a getter, because of [the way we forward them to the label/legend component](https://github.com/hashicorp/design-system/pull/425/files#diff-61b99252eb7c03439d9402878520f0bfe18b8ecdf6e9d5c71188cdc0035ef168R9).

I can convert them from `@tracked` to only having a getter, e.g.
```
  get isRequired() {
    let { isRequired } = this.args;
    return isRequired;
  }
```

didoo replied:
Mmm, ok now I see why (that was the part I missed: the fact that when you hash them, they need to be `this.something` and not `this.args.something`, while I was thinking "in React" where automatically a prop is automatically exposed)

Personally, I have a subtle preference for the getter, probably in an ever simplified way:
```
// I would add a comment here explaining why we need to do this strange thing
get isRequired() {
    return this.args.isRequired;
  }
```

didoo commented:
> @didoo While it's harder to accomplish in Figma (and would require a detach of the component), the indicator wrapping exactly to that point is such an edge case that I don't think we should be adjusting our whole approach around it. Personally, the inline approach feels like it provides a more natural reading/scanning experience.

FYI this is how it would work in a browser:
<img width="206" alt="screenshot_1617" src="https://user-images.githubusercontent.com/686239/177556492-013494e7-551c-4c28-8b1c-e5568fa87420.png">
so the case you mentioned would be only in Figma.

> [question] Could you move the 4px margin to the text and not the badge so that the gap on the left is avoided?

@alex-ju I _think_ this would require to move the mark outside the `<label/legend>` and use a wrapper container, right? in that case do you see any impact from an accessibility standpoint? @MelSumner @alex-ju should the "required/optional" mark live _inside_ the `<label/legend>`, _outside_ of it, or it's irrelevant?

alex-ju commented on packages/components/addon/components/hds/form/label/index.hbs:
```diff
@@ -1,3 +1,5 @@
 <label class={{this.classNames}} for={{@controlId}} ...attributes>
   {{yield}}
+  {{#if @isRequired}}<Hds::Badge @size="small" @text="Required" />{{/if}}
```
My assumption was that we'd want the 'Required' badge to be a default badge, regardless of what that would look like. Have updated now to explicitly declare it as 'neutral'

alex-ju commented on packages/components/addon/components/hds/form/legend/index.hbs:
```diff
@@ -1,3 +1,5 @@
 <legend class={{this.classNames}} ...attributes>
   {{yield}}
+  {{#if @isRequired}}<Hds::Badge @size="small" @text="Required" />{{/if}}
+  {{#if @isOptional}}<span class="hds-form-legend__optional-indicator">(Optional)</span>{{/if}}
```
Have abstracted these into `required-indicator` and `optional-indicator` within the `forms` directory.

alex-ju commented on packages/components/app/styles/components/badge.scss:
```diff
@@ -144,3 +144,12 @@ $size-props: (
+
+.hds-form-label,
+.hds-form-legend {
+  .hds-badge {
```
Agreed – have moved these styles into their own files as part of the abstraction suggested above.

alex-ju commented:
> > I omitted the toggle component as I've not seen a reference in design and don't think it makes sense for the component to have it
> 
> @alex-ju probably for the same reason for which we re-introduced the "group" variant for the toggle, and we showcase it as multiple toggles even if likely is not a real use case, I would be inclined to add it anyway (maybe a user need to turn on a "toggle" for some reasons, like accepting the risk of an operation, or explicitly approving something)

@didoo yes, I can rebase to main and add it for the 'group' variant, but I'd like to check with Christian first so we keep in line with design.

@cveigt we're planning to introduce the `isRequired` and `isOptional` indicators at the field level for text-input, textarea and select and at the group level for the checkboxes and radio. what are your thoughts on these indicators for the toggle? shall we have them as an option, and if so at the field or group level? or both?

alex-ju commented:
@didoo, @MelSumner I think this is ready for another review. I added a few commits on top of what you reviewed already and rebased to `main` to get [the new additions to the Toggle component](https://github.com/hashicorp/design-system/pull/421).
- Swapped from `@tracked` to getters for `isRequired` and `isOptional`, setting a default of `false`
- Explicitly defined the 'Require' badge color
- Refactored to abstract optional/required indicators, as suggested, removing some styling duplication
- Generalised parameter description then aligned it with test names
- Updated the structured content examples and added more examples for indicators
- Set the `required` attribute on form controls when `isRequired` is `true`
- Add `isRequired` and `isOptional` to `Form::Toggle` to bring it in line with the checkbox component.

didoo commented on packages/components/tests/dummy/app/templates/components/form/checkbox.hbs:
```diff
@@ -183,6 +183,18 @@
+    <dd>
+      <p>Appends an indicator to the legend text and sets the
+        <code class="dummy-code">required</code>
+        attribute when user input is required.</p>
```
I had to read the text multiple times to understand the role of the attribute. Maybe being mode specific makes it more obvious what happens behind the scenes, what receives that attribute.

```suggestion
        attribute on the controls when user input is required.</p>
```

didoo commented on packages/components/tests/dummy/app/templates/components/form/radio.hbs:
```diff
@@ -183,6 +183,18 @@
+    <dd>
+      <p>Appends an indicator to the legend text and sets the
+        <code class="dummy-code">required</code>
+        attribute when user input is required.</p>
```
```suggestion
        attribute on the controls when user input is required.</p>
```

didoo commented on packages/components/tests/dummy/app/templates/components/form/select.hbs:
```diff
@@ -111,6 +111,18 @@
+    <dd>
+      <p>Appends an indicator to the label text and sets the
+        <code class="dummy-code">required</code>
+        attribute when user input is required.</p>
```
```suggestion
        attribute on the control when user input is required.</p>
```

didoo commented on packages/components/tests/dummy/app/templates/components/form/text-input.hbs:
```diff
@@ -129,6 +129,18 @@
+    <dd>
+      <p>Appends an indicator to the label text and sets the
+        <code class="dummy-code">required</code>
+        attribute when user input is required.</p>
```
```suggestion
        attribute on the control when user input is required.</p>
```

didoo commented on packages/components/tests/dummy/app/templates/components/form/textarea.hbs:
```diff
@@ -100,6 +100,18 @@
+    <dd>
+      <p>Appends an indicator to the label text and sets the
+        <code class="dummy-code">required</code>
+        attribute when user input is required.</p>
```
```suggestion
        attribute on the control when user input is required.</p>
```

didoo commented on packages/components/tests/dummy/app/templates/components/form/toggle.hbs:
```diff
@@ -177,6 +177,18 @@
+    <dd>
+      <p>Appends an indicator to the legend text and sets the
+        <code class="dummy-code">required</code>
+        attribute when user input is required.</p>
```
```suggestion
        attribute on the controls when user input is required.</p>
```

didoo commented on packages/components/tests/integration/components/hds/form/checkbox/group-test.js:
```diff
@@ -66,6 +66,36 @@ module('Integration | Component | hds/form/checkbox/group', function (hooks) {
+    );
+    assert.dom('legend .hds-form-required-indicator').exists();
+    assert.dom('legend .hds-form-required-indicator').hasText('Required');
+    assert.dom('input[required]').exists();
```
maybe I would use something more explicit about what we're testing
```suggestion
    assert.dom('input').hasAttribute('required');
```

didoo commented on packages/components/tests/integration/components/hds/form/radio/group-test.js:
```diff
@@ -66,6 +66,36 @@ module('Integration | Component | hds/form/radio/group', function (hooks) {
+    );
+    assert.dom('legend .hds-form-required-indicator').exists();
+    assert.dom('legend .hds-form-required-indicator').hasText('Required');
+    assert.dom('input[required]').exists();
```
```suggestion
    assert.dom('input').hasAttribute('required');
```

didoo commented on packages/components/tests/integration/components/hds/form/select/field-test.js:
```diff
@@ -94,6 +94,30 @@ module('Integration | Component | hds/form/select/field', function (hooks) {
+    );
+    assert.dom('label .hds-form-required-indicator').exists();
+    assert.dom('label .hds-form-required-indicator').hasText('Required');
+    assert.dom('select[required]').exists();
```
```suggestion
    assert.dom('select').hasAttribute('required');
```

didoo commented on packages/components/tests/integration/components/hds/form/text-input/field-test.js:
```diff
@@ -101,6 +101,30 @@ module('Integration | Component | hds/form/text-input/field', function (hooks) {
+    );
+    assert.dom('label .hds-form-required-indicator').exists();
+    assert.dom('label .hds-form-required-indicator').hasText('Required');
+    assert.dom('input[required]').exists();
```
```suggestion
    assert.dom('input').hasAttribute('required');
```

didoo commented on packages/components/tests/integration/components/hds/form/textarea/field-test.js:
```diff
@@ -82,6 +82,31 @@ module('Integration | Component | hds/form/textarea/field', function (hooks) {
+    );
+    assert.dom('label .hds-form-required-indicator').exists();
+    assert.dom('label .hds-form-required-indicator').hasText('Required');
+    assert.dom('textarea[required]').exists();
```
```suggestion
    assert.dom('textarea').hasAttribute('required');
```

didoo commented on packages/components/tests/integration/components/hds/form/toggle/group-test.js:
```diff
@@ -66,6 +66,36 @@ module('Integration | Component | hds/form/toggle/group', function (hooks) {
+    );
+    assert.dom('legend .hds-form-required-indicator').exists();
+    assert.dom('legend .hds-form-required-indicator').hasText('Required');
+    assert.dom('input[required]').exists();
```
```suggestion
    assert.dom('input').hasAttribute('required');
```

didoo commented on packages/components/app/styles/components/form/optional-indicator.scss:
```diff
@@ -0,0 +1,10 @@
+// properties within each class are sorted alphabetically
+//
+
+.hds-form-optional-indicator {
+  color: var(--token-color-foreground-faint);
+  font-weight: var(--token-typography-font-weight-regular);
+}
```
I _think_ we've lost the styling that sets the sizing of the element (probably just the `line-height`? I remember I checked the size, but maybe I misremember).

Reason to have it, is that in some conditions the alignment may appear off:
<img width="1792" alt="screenshot_1621" src="https://user-images.githubusercontent.com/686239/177790695-39f9eaf0-5ba8-4f0b-a578-dfe07c94b3c2.png">

didoo reviewed (commented):
Nothing outstanding, I've left a bunch of comments, but I would also discuss a few things in person before approving:
- checkbox/radio/toggle fields: should they have `@isRequired/@isOptional` fields for consistency (and potentially some use cases, especially checkbox and toggle?
- `@isRequired` as extra element also at "control:base" level? (99% no, but want to hear your thoughts on this)
- merging the "required" and "optional" components in a single one: pros vs cons

I'll set up a meeting for later (or tomorrow)

cveigt commented:
Small detail: There is a slight misalignment between the label and the badge.
This is probably not the solution, but I was able to "fix it" by removing `vertical-align` from the badge.

<img width="1341" alt="Screen Shot 2022-07-07 at 1 07 05 PM" src="https://user-images.githubusercontent.com/2142312/177862473-d7578799-b96a-4af4-a377-7a2e9828bc8d.png">

<img width="1600" alt="Screen Shot 2022-07-07 at 1 11 27 PM" src="https://user-images.githubusercontent.com/2142312/177863185-0614d893-8801-4dc2-afd4-9dfeba56e9c2.png">

alex-ju commented:
> This is probably not the solution, but I was able to "fix it" by removing `vertical-align` from the badge.

@cveigt yes, there is a misalignment, indeed, and removing the `vertical-align` property is the solution for this case (a simple badge in a label), but it will [break the alignment for badges with icons](https://github.com/hashicorp/design-system/pull/406) (including the case where a badge with icon is placed in a label) 😭  – I'll give this some more thought.

alex-ju commented on packages/components/app/styles/components/form/optional-indicator.scss:
```diff
@@ -6,5 +6,7 @@
 .hds-form-optional-indicator {
   color: var(--token-color-foreground-faint);
+  line-height: 1.23077; // 16px
+  font-size: 0.8125rem; // 13px
```
Makes sense, will update – I borrowed the ones from the badge, thinking we want to align the two.

cveigt commented:
@alex-ju back to the required/optional indicators. We reviewed this today and will have them at the group level, but not the field, which I think is how you have them now.

alex-ju commented:
@MelSumner I hope I addressed all your comments on this PR as well – would you like to have another look?

alex-ju commented:
I'll carry on and merge this to avoid potential conflicts in other branches, but let me know if anything and I'll open follow-up PRs
